src/coreclr/vm/gchelpers.cpp

```diff
@@ -183,6 +183,126 @@ inline void CheckObjectSize(size_t alloc_size)
+    }
+}
+
+inline Object* Alloc(ee_alloc_context* pEEAllocContext, size_t size, GC_ALLOC_FLAGS flags)
+{
+    CONTRACTL {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE; // returns an objref without pinning it => cooperative
+    } CONTRACTL_END;
+
+    Object* retVal = nullptr;
+    gc_alloc_context* pAllocContext = &pEEAllocContext->gc_alloc_context;
+    // the GC is aligning the allocated objects
+    // see gc_heap::allocate() for more details
+    size_t alignment = sizeof(uintptr_t) - 1;
+    size_t allocatedBytes = (size + alignment) & ~alignment;
+
+    size_t samplingBudget = (size_t)(pEEAllocContext->alloc_sampling - pAllocContext->alloc_ptr);
+    size_t availableSpace = (size_t)(pAllocContext->alloc_limit - pAllocContext->alloc_ptr);
+    auto pCurrentThread = GetThread();
+
+    bool isSampled = false;
+    if (ETW_TRACING_CATEGORY_ENABLED(MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_DOTNET_Context,
+                                     TRACE_LEVEL_INFORMATION,
+                                     CLR_ALLOCATIONSAMPLING_KEYWORD))
+    {
+        if (flags & GC_ALLOC_USER_OLD_HEAP)
+        {
+            // if sampling is on, decide if this object should be sampled.
+            // the provided allocation context fields are not used for LOH/POH allocations
+            // (only its alloc_bytes_uoh field will be updated),
+            // so get a random size in the distribution and if it is less than the size of the object
+            // then this object should be sampled
+            samplingBudget = ee_alloc_context::ComputeGeometricRandom(pCurrentThread->GetRandom());
+            isSampled = (samplingBudget < allocatedBytes);
+        }
+        else
+        {
+            // Check to see if the allocated object overlaps a sampled byte
+            // in this AC. This happens when both:
+            // 1) The AC contains a sampled byte (alloc_sampling < alloc_limit)
+            // 2) The object is large enough to overlap it (allocatedBytes > samplingBudget)
+            //
+            // Note that the AC could have no remaining space for allocations (alloc_ptr =
+            // alloc_limit = alloc_sampling). When a thread hasn't done any SOH allocations
+            // yet it also starts in an empty state where alloc_ptr = alloc_limit =
+            // alloc_sampling = nullptr. The (1) check handles both of these situations
+            // properly as an empty AC can not have a sampled byte inside of it.
+            isSampled =
+                (pEEAllocContext->alloc_sampling < pAllocContext->alloc_limit) &&
+                (allocatedBytes > samplingBudget);
+
+            // if the object overflows the AC, we need to sample the remaining bytes
+            // the sampling budget only included at most the bytes inside the AC
+            if (allocatedBytes > availableSpace && !isSampled)
+            {
+                samplingBudget = ee_alloc_context::ComputeGeometricRandom(pCurrentThread->GetRandom());
```

Comment:
There is a bug here that is contributing (probably causing entirely) the skew you are seeing in the test results. For an object we sample that extends past the end of the AC the byte we sampled was at offset `availableSpace + ComputeGeometricRandom()`. Right now we are reporting the offset not including `availableSpace` which is biasing the reported offsets to be lower than they really were.

Reply:
I'm thinking about changing what is provided by AllocationSampled payload: instead of the budgetOffset, it would be more practical to directly send the remainder that is used by the listener.
```
<-- availableSpace ----------------> <-- samplingBudget --> <-- remainder -->
| ptr                         limit |                      |xxxxxxxxxxxxxxxx|
|/_________________________________\|______________________|xxxxxxxxxxxxxxxx|
<------------------------------ allocateBytes ------------------------------>
```
That would ease the work of the listener because the remainder would be immediately available.
Otherwise, in that particular case, the budgetOffset needs to be (availableSpace + samplingBudget) just because we want to have the same code in the listener: remainder = objectSize - budgetOffset.

As a side note, if we think about it, the first part availableSpace was already "tried" but the last ComputeGeometricRandom() returned a value greater than size(AC). The size of the AC is probably much larger than availableSpace (because some smaller objects were allocated before the current one). Am I wrong by saying that the bytes in availableSpace were already "checked"? Do we need to take them into account? From my limited mathematical understanding, the bytes to check should be allocatedByte - size(AC) and not allocatedBytes - availableSpace. However, we don't know size(AC) so it is an approximation.

Reply:
> instead of the budgetOffset, it would be more practical to directly send the remainder that is used by the listener.

We could, but it seems unnecessary. Any piece of code that has two of the values (size, offset, remainder) can always calculate the 3rd one with a single addition or subtraction. 

I included a fix for this issue in the commit I pushed an hour ago that IMO looks just fine. Take a look and see what you think.

> Am I wrong by saying that the bytes in availableSpace were already "checked"?

Thats correct, the bytes in availableSpace already had bernoulli trials performed which were determined to be failed trails.

> Do we need to take them into account? From my limited mathematical understanding, the bytes to check should be allocatedByte - size(AC) and not allocatedBytes - availableSpace. However, we don't know size(AC) so it is an approximation.

For an object that extends past the end of the AC, we are calculating whether there is a successful trial somewhere in the remaining bytes of the object. Since we've already done trials for `availableSpace` number of bytes that leaves `allocatedBytes - availableSpace` bytes that still need to be tested. Not sure if the commit I pushed earlier this afternoon makes that any clearer.